Add EMR Serverless Operators and Hooks

syedahsn opened the pull request:
Closes: #20215

Define new operators, as well as necessary hooks and sensors, to run EMR Serverless jobs. System test for EMR Serverless will be coming in a separate PR.
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

syedahsn committed:
Hooks and Operators for EMR Serverless

syedahsn committed:
Remove unused imports

syedahsn committed:
Updated work for Dennis

syedahsn committed:
Fix bugs with starting operators

syedahsn committed:
Add sensor for emr jobs and applications

ferruzzi committed:
Linting, Added Missing Docs, and CR Fixes

ferruzzi committed:
Fix Fake Waiter

Co-Author: https://github.com/dacort

syedahsn committed:
EMR dags and docs CR

syedahsn committed:
CR comments rev 2

syedahsn committed:
rev 8 changes suggested by Niko

syedahsn committed:
Fix typos and documentation for rev 8 comments

syedahsn committed:
Move waiter logic to hook, improve sensor functionality to report failures

syedahsn committed:
Move wait function to work only when needed

syedahsn committed:
Tests for emr serverless

syedahsn committed:
Update emr serverless release label to latest version

syedahsn committed:
Make changes suggested by Niko

syedahsn committed:
Mock waiter function call - improve emr serverless hook tests

syedahsn committed:
Add support for arbitrary parameters to StartJobOperator

syedahsn committed:
Change waiter to make check_interval_seconds and countdown be passed as params. Add test to check timeout exception

syedahsn committed:
Remove duplicate test variables

syedahsn committed:
Remove service.json

potiuk commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -16,18 +16,24 @@
 if TYPE_CHECKING:
     from airflow.utils.context import Context
 
+if sys.version_info >= (3, 8):
```
I tihnk we are already fine with our util class (as in other providers) since we moved to 2.2+

josh-fell replied:
+1 You can import `cached_property` from [`airflow.compat.functools`](https://github.com/apache/airflow/blob/main/airflow/compat/functools.py).

syedahsn replied:
Sounds good, I'll make the change

josh-fell commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -90,6 +91,78 @@ def create_job_flow(self, job_flow_overrides: Dict[str, Any]) -> Dict[str, Any]:
+        :param countdown: Total amount of time the waiter should wait for the desired state
+        before timing out (in seconds). Defaults to 25 * 60 seconds.
+        :param check_interval_seconds: Number of seconds waiter should wait before attempting
+        to retry get_state_callable. Defaults to 60 seconds.
```
```suggestion
        :param failure_states: A set of states which indicate failure and should throw an
            exception if any are reached before the desired_state
        :param object_type: Used for the reporting string. What are you waiting for? (application, job, etc)
        :param action: Used for the reporting string. What action are you waiting for? (created, deleted, etc)
        :param countdown: Total amount of time the waiter should wait for the desired state
            before timing out (in seconds). Defaults to 25 * 60 seconds.
        :param check_interval_seconds: Number of seconds waiter should wait before attempting
            to retry get_state_callable. Defaults to 60 seconds.
```
Would you mind add some indentation? The docstrings typically have a hanging indent in parameter descriptions on lines beyond the first for readability.

josh-fell commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -90,6 +91,78 @@ def create_job_flow(self, job_flow_overrides: Dict[str, Any]) -> Dict[str, Any]:
+                raise AirflowException(f'{object_type.title()} reached failure state {state}.')
+            if countdown >= check_interval_seconds:
+                countdown -= check_interval_seconds
+                print(f'Waiting for {object_type.lower()} to be {action.lower()}.')
```
Might be better to use `logging` instead of `print`? Again, just another typical practice.

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -15,15 +15,20 @@
+if sys.version_info >= (3, 8):
+    from functools import cached_property
+else:
+    from cached_property import cached_property
```
```suggestion
from airflow.compat.functools import cached_property
```
Same idea here as above. You can use the native compat util for `cached_property`

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+
+    :param application_id: application_id to check the state of
+    :param job_run_id: job_run_id to check the state of
+    :param target_states: a set of states to wait for, defaults to SUCCESS_STATES
```
```suggestion
    :param target_states: a set of states to wait for, defaults to 'SUCCESS'
```
Maybe it makes sense to use the actual value being checked rather than the variable in the docstring. I was thinking users reading the Python API docs won't have to look at code or other places in the docs to figure out what `SUCCESS_STATES` is. WDYT?

vincbeck replied:
Agree

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+    template_fields: Sequence[str] = ('application_id',)
+
+    INTERMEDIATE_STATES = {'CREATING', 'STARTING', 'STOPPING'}
+    # TODO:  Question: Do these states indicate failure?
```
@vincbeck Is this something you can answer?

syedahsn replied:
This was a question posed to one of the internal reviewers. I'll remove it from here.

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+    @cached_property
+    def hook(self) -> EmrServerlessHook:
+        """Create and return an EmrServerlessHook"""
+        return EmrServerlessHook()
```
```suggestion
        return EmrServerlessHook(aws_conn_id=self.aws_conn_id)
```
So users can control the connection ID used to build the hook object. Unless this should be something like  use `aws_conn_id` if provided or `emr_conn_id` otherwise?

vincbeck replied:
Agree, we should pass `aws_conn_id`

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+    @cached_property
+    def hook(self) -> EmrServerlessHook:
+        """Create and return an EmrServerlessHook"""
+        return EmrServerlessHook()
```
```suggestion
        return EmrServerlessHook(aws_conn_id=self.aws_conn_id)
```

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+        **kwargs: Any,
+    ) -> None:
+        self.aws_conn_id = aws_conn_id
+        self.emr_conn_id = emr_conn_id
```
I could easily be missing it, but where is `emr_conn_id` being used in this sensor and other sensors in this file?

vincbeck replied:
I agree. It seems `emr_conn_id` is not used. You can just remove it

josh-fell commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+        :ref:`howto/sensor:EmrServerlessApplicationSensor`
+
+    :param application_id: application_id to check the state of
+    :param target_states: a set of states to wait for, defaults to SUCCESS_STATES
```
```suggestion
    :param target_states: a set of states to wait for, defaults to {'STARTED', 'CREATED'}
```
Same idea about having the values in the docstring. Up to you.

vincbeck commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -90,6 +91,78 @@ def create_job_flow(self, job_flow_overrides: Dict[str, Any]) -> Dict[str, Any]:
+                state = self.get_state(get_state_callable(**get_state_args), parse_response)
+            else:
+                message = f'{object_type.title()} still not {action.lower()} after the allocated time limit.'
+                print(message)
```
Same here I guess :)

vincbeck commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+        super().__init__(**kwargs)
+
+    def poke(self, context: 'Context') -> bool:
+        state = None
```
I dont think this line is needed

vincbeck commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -111,6 +116,152 @@ def failure_message_from_response(response: Dict[str, Any]) -> Optional[str]:
+
+        try:
+            response = self.hook.conn.get_job_run(applicationId=self.application_id, jobRunId=self.job_run_id)
+        except Exception:
```
Overall question. Do we want or is there any value converting the `Exception` to `AirflowException`? I can see there is a lot of operators which do not do that. If there is no value, I would rather not doing the conversion just for the sake of having a code cleaner

syedahsn replied:
I went with `Exception` just as a catchall for any type of exception that might be thrown by the boto3 API. I'm not sure if there's an advantage to throwing an `AirflowException` instead of a generic `Exception`. Only thing I can think of is that it would make it clear that the exception was coming from Airflow, rather than any underlying API calls (i.e. boto3) that might have been made. But that can also be understood from the message associated with the `Exception`.

josh-fell replied:
FWIW I think `AirflowException` is used too broadly in the code base. IMO `AirflowException`, and its subclasses, are really related to Airflow orchestration activity rather than activity done by what Airflow is _orchestrating_. It might make more sense to drop the `try/except` and bubble up whatever API call errors that might be thrown directly. Up to you though.

vincbeck commented on docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:
```diff
@@ -0,0 +1,111 @@
+------------------
+
+.. include:: _partials/prerequisite_tasks.rst
+
```
```suggestion
Operators
---------
```

vincbeck commented on docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:
```diff
@@ -0,0 +1,111 @@
+.. _howto/operator:EmrServerlessCreateApplicationOperator:
+
+Create an EMR Serverless Application
+------------------------------------
```
```suggestion
==========================
```

syedahsn committed:
Make changes requested by reviewers:
- Import cached_property directly functools
- Fix indentation in doc string
- Define default values in doc string explicitly
- Remove unnecessary references to emr_conn_id
- Replace print messages with log messages

josh-fell commented on docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:
```diff
@@ -0,0 +1,112 @@
+.. _howto/operator:EmrServerlessCreateApplicationOperator:
+
+Create an EMR Serverless Application
+==========================
```
This underline needs to be at least the length of the text it's underlining. I imagine this might be a Build Docs failure for you.

josh-fell commented on airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -412,3 +414,254 @@ def execute(self, context: 'Context') -> None:
+        self.log.info('Starting job on Application: %s', self.application_id)
+
+        app_state = self.hook.conn.get_application(applicationId=self.application_id)['application']['state']
+        if app_state not in {'CREATED', 'STARTED'}:
```
```suggestion
        if app_state not in EmrServerlessApplicationSensor.SUCCESS_STATES:
```
Small suggestion to use the attrs where you can like you have in other places. Up to you though (assuming it's functionally correct as well).

josh-fell reviewed (approved):
Some minor suggestions but overall LGTM (after fixing the tests of course).

syedahsn committed:
Removed redundant try/except around calls to boto3 API
 - Fixed minor formatting in emr_serverless.rst

ferruzzi commented on airflow/providers/amazon/aws/sensors/emr.py:
```diff
@@ -219,10 +216,7 @@ def __init__(
     def poke(self, context: 'Context') -> bool:
         state = None
 
-        try:
```
certainly makes to code cleaner.  :+1:

josh-fell commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -90,6 +91,79 @@ def create_job_flow(self, job_flow_overrides: Dict[str, Any]) -> Dict[str, Any]:
+                raise AirflowException(f'{object_type.title()} reached failure state {state}.')
+            if countdown >= check_interval_seconds:
+                countdown -= check_interval_seconds
+                print(f'Waiting for {object_type.lower()} to be {action.lower()}.')
```
FYI - A couple of `print()` statements here.

vincbeck commented on docs/apache-airflow-providers-amazon/operators/emr_serverless.rst:
```diff
@@ -0,0 +1,112 @@
+------------------
+
+.. include:: _partials/prerequisite_tasks.rst
+Operators
+---------
```
```suggestion

Operators
---------

```

eladkal commented on airflow/providers/amazon/aws/example_dags/example_emr_serverless.py:
```diff
@@ -0,0 +1,97 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
new example dags should be according to AIP-47
Example: https://github.com/apache/airflow/pull/25205

vincbeck replied:
That's a good point but the plan was to do it in two iterations since this PR is already big enough. Once this PR is merged, @syedahsn will work on converting this sample dag to system test. If you strongly disagree, I guess the two can be done in this PR but it will just make it bigger and harder to review

syedahsn committed:
Forgot to remove some print statements
 - Minor formatting in emr_serverless.rst

syedahsn committed:
Add template fields to EmrServerlessStartJobOperator
